src/decoder/stateful.rs

```diff
@@ -0,0 +1,169 @@
+// Copyright 2023 The ChromiumOS Authors
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+use thiserror::Error;
+use zerocopy::AsBytes;
+
+use crate::decoder::DecodedHandle;
+use crate::decoder::StreamInfo;
+use crate::DecodedFormat;
+use crate::Resolution;
+
+use std::sync::Arc;
+
+pub enum DecodeError {
+    InvalidState,
+    InvalidData,
+}
+
+pub enum ConfigError {
+    InvalidState,
+    NotSupported,
+    InvalidConfig,
+}
+
+/// Error returned by stateful backend methods.
+#[derive(Error, Debug)]
+pub enum StatefulBackendError {
+    #[error("not enough resources to proceed with the operation now")]
+    OutOfResources,
+    #[error("this format is not supported")]
+    UnsupportedFormat,
+    #[error(transparent)]
+    Other(#[from] anyhow::Error),
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum EncodedFormat {
+    AV1,
+    H264,
+    H265,
+    VP8,
+    VP9,
+}
+
+/// Common trait shared by all stateful video decoder backends, providing codec-independent
+/// methods.
+pub trait StatefulDecoderBackend<Codec: StatefulCodec> {
+    /// The type that the backend returns as a result of a decode operation.
+    /// This will usually be some backend-specific type with a resource and a
+    /// resource pool so that said buffer can be reused for another decode
+    /// operation when it goes out of scope.
+    type Handle: DecodedHandle;
+
+    /// Returns the current decoding parameters, as parsed from the stream.
+    fn stream_info(&self) -> Option<&StreamInfo>;
+}
+
+pub trait StatefulCodec {
+    /// State that needs to be kept during a decoding operation, typed by backend.
+    type DecoderState<B: StatefulDecoderBackend<Self>>;
+}
+
+#[derive(Copy, Clone, Debug)]
+
+pub enum ColorGamut {
+    Bt709,
+    Bt470bg,
+    Smpte170m,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum ColorTransfer {
+    Bt709,
+    Smpte170m,
+    Iec61966_2_1,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum ColorMatrix {
+    Rgb,
+    Bt709,
+    Bt470bg,
+    Smpte170m,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct VideoColorSpace {
+    primaries: ColorGamut,
+    transfer: ColorTransfer,
+    matrix: ColorMatrix,
+}
+
+#[derive(Clone, Debug)]
+pub struct EncodedVideoChunk {
+    data: Arc<Vec<u8>>,
+}
+
+pub struct VideoFrame {
+    format: DecodedFormat,
+    coded_resolution: Resolution,
+    display_resolution: Resolution,
+    duration: usize,
+    timestamp: usize,
+    color_space: VideoColorSpace,
+    buffer: Option<Box<dyn AsBytes>>,
+}
+
+pub enum StatefulDecoderEvent {
```

Comment:
I suspect you can reuse `DecoderEvent` (and the `DecodedHandle` trait) as-is for the stateful decoder. It will align the interfaces and make the code simpler somehow.